src/connectivity/mod.rs

```diff
@@ -225,17 +227,41 @@ pub fn digraph_find_cycle(
     EdgeList { edges: cycle }
 }
 
-/// Find the number of weakly connected components in a DAG.
+/// Find the number of weakly connected components in a directed graph
 ///
 /// :param PyDiGraph graph: The graph to find the number of weakly connected
 ///     components on
 ///
-/// :returns: The number of weakly connected components in the DAG
+/// :returns: The number of weakly connected components in the graph
 /// :rtype: int
 #[pyfunction]
 #[pyo3(text_signature = "(graph, /)")]
 fn number_weakly_connected_components(graph: &digraph::PyDiGraph) -> usize {
-    algo::connected_components(graph)
+    if graph.graph.node_count() == 0 {
+        return 0;
+    }
+    let mut vertex_sets = UnionFind::new(graph.node_bound());
+    let first = graph
+        .graph
+        .node_indices()
+        .next()
+        .map(|a| graph.graph.to_index(a))
+        .unwrap();
+    for ix in 0..graph.node_bound() {
+        if !graph.graph.contains_node(NodeIndex::new(ix)) {
+            vertex_sets.union(first, ix);
+        }
+    }
+    for edge in graph.graph.edge_references() {
+        let (a, b) = (edge.source(), edge.target());
+
+        // union the two vertices of the edge
+        vertex_sets.union(graph.graph.to_index(a), graph.graph.to_index(b));
+    }
+    let mut labels = vertex_sets.into_labeling();
+    labels.sort_unstable();
+    labels.dedup();
+    labels.len()
```

Comment:
This works, but connecting the first node with the removed nodes is a bit hacky. Also, we can avoid the sort at the end.

A simpler algorithm is to subtract one from the number of components every time we do a union. If we initialize `weak_components` as the number of non-removed, we can tweak the existing code to get the answer:

```rust
for edge in graph.graph.edge_references() {
        let (u, v) = (edge.source(), edge.target());
        if vertex_sets.union(graph.graph.to_index(u), graph.graph.to_index(v)){
            weak_components -= 1;
        }
    }
```

The answer is `weak_components`

Reply:
Thanks, this is much simpler! Done in 168f0d1